supervisor/utils/systemd_journal.py

```diff
@@ -0,0 +1,112 @@
+        "_HOSTNAME",
+        "SYSLOG_IDENTIFIER",
+        "_PID",
+        "MESSAGE",
+    ]
+)
+def journal_verbose_formatter(entries: dict[str, str]) -> str:
+    """Format parsed journal entries to a journalctl-like format."""
+    ts = datetime.fromtimestamp(
+        int(entries["__REALTIME_TIMESTAMP"]) / 1e6, UTC
+    ).isoformat(sep=" ", timespec="milliseconds")
+    ts = ts[: ts.index(".") + 4]  # strip TZ offset
+
+    identifier = (
+        f"{entries["SYSLOG_IDENTIFIER"]}[{entries["_PID"]}]"
+        if "_PID" in entries
+        else entries["SYSLOG_IDENTIFIER"]
+    )
+
+    return f"{ts} {entries["_HOSTNAME"]} {identifier}: " f"{entries["MESSAGE"]}"
+
+
+async def journal_logs_reader(
+    journal_logs: ClientResponse,
+    log_formatter: LogFormatter = LogFormatter.PLAIN,
+) -> AsyncGenerator[str, None]:
+    """Read logs from systemd journal line by line, formatted using the given formatter."""
+    match log_formatter:
+        case LogFormatter.PLAIN:
+            formatter_ = journal_plain_formatter
+        case LogFormatter.VERBOSE:
+            formatter_ = journal_verbose_formatter
+        case _:
+            raise ValueError(f"Unknown log format: {log_formatter}")
+
+    async with journal_logs as resp:
+        entries: dict[str, str] = {}
+        while not resp.content.at_eof():
+            entry = await resp.content.readuntil(b"\n")
+            line = entry
+            # newline means end of message:
+            if line == b"\n":
+                yield formatter_(entries)
+                continue
+
+            # Journal fields consisting only of valid non-control UTF-8 codepoints
+            # are serialized as they are (i.e. the field name, followed by '=',
+            # followed by field data), followed by a newline as separator to the next
+            # field. Note that fields containing newlines cannot be formatted like
+            # this. Non-control UTF-8 codepoints are the codepoints with value at or
+            # above 32 (' '), or equal to 9 (TAB).
+            idx = line.find(b"=")
+            if idx != -1:
+                # name=value\n
+                name = line[:idx].decode("utf-8")
+                data = line[idx + 1 : -1]  # strip \n
+            else:
+                # Other journal fields are serialized in a special binary safe way:
+                # field name, followed by newline
+                name = line[:-1].decode("utf-8")  # strip \n
+                # followed by a binary 64-bit little endian size value,
+                length_raw = await resp.content.readexactly(8)
+                length = int.from_bytes(length_raw, byteorder="little")
+                # followed by the binary field data,
+                data = await resp.content.readexactly(length)
+                # followed by a newline as separator to the next field.
+                if (await resp.content.readexactly(1)) != b"\n":
+                    raise MalformedBinaryEntry(line)
```

Comment:
It seems we do lose some time due to the repeated read calls, we can safe this one:

```suggestion
                data = await resp.content.readexactly(length + 1)
                # followed by a newline as separator to the next field.
                if data[len(data-1)] != "\n":
                    raise MalformedBinaryEntry(line)
```

Now data will be too long though. Making an unconditional copy using `data = data[:-1]` is bad however :cry: 

But it seems not to fully recoop the time.

We probably can also safe a copy of 

```python
            if name not in formatter_.required_fields:
                continue

            if idx != -1:
                data = line[idx + 1 : -1]  # strip \n
            else:
                data = data[:-1]
            entries[name] = data.decode("utf-8")
```

It all doesn't seem to make a massive difference in a few tests here :cry:

Reply:
> we still do ~40 reads for every line for a single journal entry, so it's not surprising it doesn't make much difference.

Right :cry:  And from tracing, it seems that awaiting for the next bytes is really what slows the whole thing down.

I guess we'd need to make this "more synchronous" if we want to speed up. But it gets more complicated as we have reads in several places.

> I'll rather duplicate the if...continue in each block for readability if you don't mind.

Hacking a bit further I came up with this:

```python
            required_fields = {s.encode('utf-8'): s for s in formatter_.required_fields}
            while not resp.content.at_eof():
                ...
                name, sep, data = line.partition(b'=')
                if not sep:
                    # followed by a binary 64-bit little endian size value,
                    length_raw = await resp.content.readexactly(8)
                    length = int.from_bytes(length_raw, byteorder="little")
                    # followed by the binary field data,
                    data = await resp.content.readexactly(length + 1)
                    # followed by a newline as separator to the next field.
                    if data[len(data-1)] != "\n":
                        raise MalformedBinaryEntry(line)

                name = required_fields.get(name)
                if not name:
                    continue

                if not sep:
                    data = data[:-1]

                entries[name] = data.decode("utf-8")
```

My goal was to safe decoding name for each entry, and make less copy. I am not sure if `line.partition()` really leads to less copy, but it kinda ended up kinda neat code :sweat_smile:  But all didn't really make it faster :see_no_evil:  So not sure if the `required_fields` look up is really worth it.